File: guild/src/main/java/com/os/operando/guild/Septet.java
package com.os.operando.guild;

/**
 * A tuple of seven elements.
 *
 * @param <F>  first element type
 * @param <S>  second element type
 * @param <T>  third element type
 * @param <FO> fourth element type
 * @param <FI> five element type
 * @param <SI> six element type
 * @param <SE> seven element type
 */
public class Septet<F, S, T, FO, FI, SI, SE> {

    private final F first;
    private final S second;
    private final T third;
    private final FO fourth;
    private final FI five;
    private final SI six;
    private final SE seven;

    private Septet(final F first, final S second, final T third, final FO fourth, final FI five, final SI six, SE seven) {
        this.first = first;
        this.second = second;
        this.third = third;
        this.fourth = fourth;
        this.five = five;
        this.six = six;
        this.seven = seven;
    }

    public static <F, S, T, FO, FI, SI, SE> Septet<F, S, T, FO, FI, SI, SE> create(F first, S second, T third, FO fourth, FI five, SI six, SE seven) {
        return new Septet<>(first, second, third, fourth, five, six, seven);
    }

    public F getFirst() {
        return first;
    }

    public S getSecond() {
        return second;
    }

    public T getThird() {
        return third;
    }

    public FO getFourth() {
        return fourth;
    }

    public FI getFive() {
        return five;
    }

    public SI getSix() {
        return six;
    }

    public SE getSeven() {
        return seven;
    }

    @Override
    public String toString() {
        return "Septet{" +
                "first=" + first +
                ", second=" + second +
                ", third=" + third +
                ", fourth=" + fourth +
                ", five=" + five +
                ", six=" + six +
                ", seven=" + seven +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Septet<?, ?, ?, ?, ?, ?, ?> septet = (Septet<?, ?, ?, ?, ?, ?, ?>) o;

        if (first != null ? !first.equals(septet.first) : septet.first != null) return false;
        if (second != null ? !second.equals(septet.second) : septet.second != null) return false;
        if (third != null ? !third.equals(septet.third) : septet.third != null) return false;
        if (fourth != null ? !fourth.equals(septet.fourth) : septet.fourth != null) return false;
        if (five != null ? !five.equals(septet.five) : septet.five != null) return false;
        if (six != null ? !six.equals(septet.six) : septet.six != null) return false;
        return seven != null ? seven.equals(septet.seven) : septet.seven == null;
    }

    @Override
    public int hashCode() {
        int result = first != null ? first.hashCode() : 0;
        result = 31 * result + (second != null ? second.hashCode() : 0);
        result = 31 * result + (third != null ? third.hashCode() : 0);
        result = 31 * result + (fourth != null ? fourth.hashCode() : 0);
        result = 31 * result + (five != null ? five.hashCode() : 0);
        result = 31 * result + (six != null ? six.hashCode() : 0);
        result = 31 * result + (seven != null ? seven.hashCode() : 0);
        return result;
    }
}

File: guild-kotlin/src/main/kotlin/com/os/operando/guild/kt/Septet.kt
package com.os.operando.guild.kt

import java.io.Serializable

/**
 * A tuple of seven elements.
 *
 * @param F  first element type
 * @param S  second element type
 * @param T  third element type
 * @param FO fourth element type
 * @param FI five element type
 * @param SI six element type
 * @param SE seven element type
 * @property first First value
 * @property second Second value
 * @property third Third value
 * @property fourth Fourth value
 * @property five Five value
 * @property six Six value
 * @property seven Seven value
 */
data class Septet<out F, out S, out T, out FO, out FI, out SI, out SE>(
        val first: F,
        val second: S,
        val third: T,
        val fourth: FO,
        val five: FI,
        val six: SI,
        val seven: SE) : Serializable {
    override fun toString(): String = "($first, $second, $third, $fourth, $five, $six, $seven)"
}

infix fun <A, B, C, D, E, F, G, H> Septet<A, B, C, D, E, F, G>.to(that: H): Octet<A, B, C, D, E, F, G, H>
        = Octet(this.first, this.second, this.third, this.fourth, this.five, this.six, this.seven, that)

fun <T> Septet<T, T, T, T, T, T, T>.toList(): List<T> = listOf(first, second, third, fourth, five, six, seven)